### PR TITLE
Add dedicated exception when try to stop self with ActorContext#stop in Akka typed.

### PR DESCRIPTION
**Pull request by He-Pin**
I explored the source code when I was reading the https://doc.akka.io/docs/akka/current/typed/guide/tutorial_1.html#the-actor-lifecycle，
and found this should be improved a little.

### CODE REVIEW

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala [hunk 1]
```diff
@@ -46,6 +46,12 @@ import scala.concurrent.duration._
             // child that was already stopped
           }
       }
+    } else if (child == self) {
```
Comment (hepin1989) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala]: This should be limited too.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala [hunk 2]
```diff
@@ -24,7 +24,7 @@ import scala.concurrent.duration._
   // lazily initialized
   private var actorLogger: OptionVal[Logger] = OptionVal.None
 
-  override def self = ActorRefAdapter(untyped.self)
+  override val self = ActorRefAdapter(untyped.self)
```
Comment (hepin1989) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala]: I changed this to `val`, we will need it anyway.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala [hunk 3]
```diff
@@ -46,6 +46,12 @@ import scala.concurrent.duration._
             // child that was already stopped
           }
       }
+    } else if (self == child) {
+      throw new IllegalArgumentException(
+        "Only direct children of an actor can be stopped through the actor context, " +
+          s"but you try to stop [$self] by passed its ActorRef to the `stop` method. " +
+          "Stopping self has to be expressed as explicitly returning a Stop Behavior " +
+          "with `Behaviors.stopped` or `Behaviors.stopped()`.")
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala]: `Behaviors.stopped` is enough, close enough for Java to

**Review by patriknw — APPROVED** [review]
LGTM, good clarification

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala [hunk 4]
```diff
@@ -46,6 +46,12 @@ import scala.concurrent.duration._
             // child that was already stopped
           }
       }
+    } else if (self == child) {
+      throw new IllegalArgumentException(
+        "Only direct children of an actor can be stopped through the actor context, " +
+          s"but you try to stop [$self] by passed its ActorRef to the `stop` method. " +
+          "Stopping self has to be expressed as explicitly returning a Stop Behavior " +
+          "with `Behaviors.stopped`.")
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala]: "but you tried to stop [$self] by passing its ActorRef to the"

**Review by johanandren — APPROVED** [review]
LGTM with one little grammar fix

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala (hunk 4 above)
Comment (hepin1989) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala]: @johanandren thanks, done.

**Review by johanandren — APPROVED** [review]
LGTM, thanks @hepin1989

**Pull request merged**